music_assistant/server/providers/siriusxm/__init__.py

```diff
@@ -0,0 +1,437 @@
+
+    async def get_stream_details(self, item_id: str) -> StreamDetails:
+        """Get StreamDetails for a radio station."""
+        # channel_id = item_id
+        # channel = self._channels_by_id[channel_id]
+
+        # playlist = await self._client.get_playlist(channel.id, use_cache=False)
+        # playlist_items = list(filter(lambda l: not l.startswith("#"), playlist.split('\n')))
+
+        # first_segment = await self._client.get_segment(playlist_items[0])
+
+        # primary_hls_root = await self._client.get_primary_hls_root()
+        # first_item = f"{primary_hls_root}/{playlist_items[0]}"
+
+        # live_channel_raw = await self._client.get_now_playing(channel)
+        # live_channel = XMLiveChannel.from_dict(live_channel_raw["moduleList"]["modules"][0])
+
+        # playlist_url = live_channel.primary_hls.url
+
+        # cache_key = f"{self.instance_id}.media_info.{channel_id}"
+        # cached_info = await self.mass.cache.get(cache_key)
+        # if cached_info:
+        #     media_info = AudioTags.parse(cached_info)
+        # else:
+        #     # parse info with ffprobe (and store in cache)
+        #     media_info = await parse_tags(self._to_generator(first_segment))
+        #     await self.mass.cache.set(cache_key, media_info.raw)
+
+        # MPEG AAC Audio (mp4a)
+        # Type: Audio
+        # Channels: 2
+        # Sample Rate: 44100
+        # Bits per sample: 32
+
+        return StreamDetails(
+            provider=self.instance_id,
+            item_id=item_id,
+            audio_format=AudioFormat(
+                sample_rate=44100,
+                channels=2,
+                bit_depth=32,
+                content_type=ContentType.M4A,
+            ),
+            # stream_title=channel.name,
+            # duration=?,
+            # data={"url": url, "format": url_details["format"]}
+            # expires=time() + 24 * 3600,
+            # size=?,
+            media_type=MediaType.RADIO,
+            # data=live_channel.primary_hls.url,
+            direct=None,
+            # can_seek=False,
+        )
+
+    async def get_playlist_items_generator(self, stream_details: StreamDetails):
+        channel_id = stream_details.item_id
+
+        if not channel_id in self._channels_by_id:
+            self.logger.warn(f"Unknown channel {channel_id}")
+            return
+
+        self.logger.info(f"Fetching playlist for channel: {channel_id}")
+        playlist = await self._client.get_playlist(channel_id, use_cache=False)
+
+        playlist_paths = list(filter(lambda l: not l.startswith("#"), playlist.split('\n')))
+        for playlist_path in playlist_paths:
+            self.logger.info(f"Getting segment: {playlist_path}")
+            yield await self._client.get_segment(playlist_path)
+
+    async def get_audio_stream(self, stream_details: StreamDetails, seek_position: int = 0) -> AsyncGenerator[bytes, None]:
+        async for chunk in self.get_playlist_items_generator(stream_details):

```

Comment:
this is not going to work because as far as I can see the "get_playlist_items_generator" returns audio file URLs and not raw audio bytes... You will have to fetch the segment through the http client and then stream the content.

It could be as simple as this:

```
from music_assistant.server.helpers.audio import get_http_stream

async for playlist_chunk_url in self.get_playlist_items_generator(stream_details):
   async for chunk in get_http_stream(
                self.mass, playlist_chunk_url, streamdetails
            ):
                yield chunk
```

Reply:
The way the sxm-client works (when you're not relying on their HTTP proxy endpoint) is that you request a playlist for a channel which returns an m3u8 file as a string. Here's a fragment of a random one I grabbed:

```
#EXTM3U
#EXT-X-VERSION:1
#EXT-X-TARGETDURATION:10
#EXT-X-MEDIA-SEQUENCE:104665
#EXT-X-ALLOW-CACHE:NO
#EXT-X-KEY:METHOD=AES-128,URI="key/1"
#EXT-X-PROGRAM-DATE-TIME:2023-12-23T14:02:23.561+00:00
#EXTINF:10,
AAC_Data/cnn/HLS_cnn_256k_v3/cnn_256k_1_122350543561_00104665_v3.aac
#EXT-X-PROGRAM-DATE-TIME:2023-12-23T14:02:33.312+00:00
#EXTINF:10,
AAC_Data/cnn/HLS_cnn_256k_v3/cnn_256k_1_122350553312_00104666_v3.aac
#EXT-X-PROGRAM-DATE-TIME:2023-12-23T14:02:43.064+00:00
#EXTINF:10,
AAC_Data/cnn/HLS_cnn_256k_v3/cnn_256k_1_122350563064_00104667_v3.aac
...etc
```

While you can get the root URL from the client (a CDN or something), the HTTP requests that you'd make require session tokens and a bunch of other stuff. Their client handles that for you. So you make a request like this:

`bytes = await sxm_client.get_segment('AAC_Data/cnn/HLS_cnn_256k_v3/cnn_256k_1_122350543561_00104665_v3.aac')`

(via https://github.com/AngellusMortis/sxm-client/blob/dd25b56d7a8a7c33399b25e72a7c2e136fb30e6a/sxm/client.py#L382)

and you get back the bytes from the http response ... where the request had all of the session-related magic taken care of behind the scenes.

Reply:
ah yes, that sounds like exactly what you need then!

Reply:
What I was trying to achieve here is to get music assistant to basically pull the next segment as needed but not try to download them all. Like can I yield the bytes of file #1, then yield the bytes of file #2 when music assistant needs them, etc.? This is what I was trying to do here. Also do you know if I would need to break these segments down into smaller chunks & yield those?

Reply:
Yeah, you just need to make sure to grab each segment, yield chunks and then grab next segment yield chunks and so on. We have some helpers in the audio.py module to help with parsing the m3u file if you want that.

and yes, you need to yield the smaller chunks otherwise it will all be loaded in memory
Basically you just reimplement the "get_segment" function from the lib but instead of returning the resp.content, you just yield chunks from the resp.iter_any or resp.iter_chunked

So something along these lines within the get_audio_stream:

```
playlist_file = get_sirius_playlist_m3u8_file
playlist_segments = parse_playlist_into_the segment_parts

for playlist_segment in playlist_segments:
    url = urllib.parse.urljoin(await self.get_hls_root(), path)
    res = await self._session.get(url, params=self._token_params())

        if res.status_code == 403:
            raise SegmentRetrievalException(
                "Received status code 403 on segment, renew session"
            )

        if res.is_error:
            self._log.warn(f"Received status code {res.status_code} on segment")
            return None

        for chunk in res.iter_any():
            yield chunk

```

You could try first by just using the entire contents from the get_segment function btw, maybe its not that big